### DISCUSSION TITLE
MQTT: sensor data not completly send out

### DISCUSSION THREAD

**Discussion by Toto1308** [arendst/Tasmota#22671] [Q&A]
Hi @all,

I use a smart meter with Tasmota and Mosquitto. First I will give all relevant data:

Tasmota version: 14.4.1
[Stromzähler_Skript_Tasmota.txt](https://github.com/user-attachments/files/18188446/Stromzahler_Skript_Tasmota.txt)

In Mosquitto following is shown for the actual config.
`Sensor={"Time":"2024-12-18T21:25:19","SM":{"76_7_0":3786.81}}`

Normaly I would expect something like this:
`sensors={"sn":{"Time":"2024-12-18T21:06:30","SM":{"1_8_0":5984848.69,"1_8_1":1001.00,"1_8_2":5983847.00,"2_8_0":2221.00,"16_7_0":11660.91,"36_7_0":4060.41,"56_7_0":3805.35,"76_7_0":3795.15,"32_7_0":233.0,"52_7_0":233.1,"72_7_0":235.1,"96_1_0":""}},"ver":1}`

So, the question now is, why are not all data send out? On the very first setup it worked. Then I tried some things with topic, full topic, device name, etc.. Since then the values are not transmitted.

What I have already done to solve the problem:
1. Reset the device
2. Reflash the device
3. Leave everything as given, except MQTT user, password and MQTT IP

Nothing seems to solve the problem. Any ideas?

If you need more informations, please give me a hint. Thank you.

Your help is really appreciated. Thank you very much in advance.

Best regards,
Thorsten

**Comment by barbudor** [comment]
I think that the tasmota discovery message is sent before the script has been able to establish the sensor structure
This doesn't happens with native drivers but would happen with sensors/energy created from Scripting
I bet the same would happen with sensors & energy from Berry

May be @gemu2015 has some idea as I pretty sure I have a discussion about that in the past

**Comment by Toto1308** [comment]
Thank you for the first assessment. I hope, there is a solution for it.

**Reply by gemu2015** [reply]
check your tele period  since the tele period cmd in the script is wrong i assume it is set to 5 minutes as default.

simply change
TelePeriod 30
to 
=>teleperiod 30

in your script

**Comment by Toto1308** [comment]
I have changed it to teleperiod 30 in my script with a restart of the device. No change...

**Reply by gemu2015** [reply]
check in console
teleperiod

did you type =>teleperiod 30 in the script ?
"=>" is important

**Comment by Toto1308** [comment]
yes, I changed it without success.

Goal is to send every time all values. Do i need the teleperiod?

**Comment by Toto1308** [comment]
In the console it states 30s for teleperiod

**Reply by gemu2015** [reply]
check what is printed in the console. values should show up every 30 seconds.
also check what is shown after a restart in the console
may be sml is not initialized properly

**Comment by Toto1308** [comment]
yes, in the console I can see every 30 seconds all values, but inbetween only one or two values are transmitted to the mqtt server.

**Comment by Toto1308** [comment]
So actually when something changes on the value side, Tasmota sends out one or two values and after 30 seconds all values. Seems to be wrong

**Reply by gemu2015** [reply]
as designed 

tele period time can be reduced down to a minimum of 10 seconds.

if you want a value to be sent immediately when received,
you must add 16 to the decimal resolution (last parameter in line)
like here
1,1-0:16.7.0*255(@1,Leistung Σ,W,16_7_0,18
(2+16 = 18)

check your script, some values are sent immediately some at tele period time

**Reply by Toto1308** [reply]
ok, thank you, i will check it

**Reply by Toto1308** [reply]
I have changed my script, please see below. 

I would like to have all data sent out at once, if something has changed. Now I have different timestamps for the values.

**Reply by gemu2015** [reply]
that is not supported directly in sml driver (only with extra scripting) !

set tele period to 10 and you will get all values at once every 10 seconds.

**Reply by Toto1308** [reply]
ok, now it works, I have changed all values from 18 back to 2 and the other ones -16. Thank you very much for your help.

**Reply by Toto1308** [reply]
One question: All values on the device HP are updated every 3 seconds. I estimate this is the rate my smartmeter sends out data. It would be great to send this directly to mqtt --> so how should the extra code look like?

**Reply by gemu2015** [reply]
first set "SM" in your meter descriptor to "*" which disables MQTT for this meter

then export every line to MQTT yourself every 3 seconds

like in this example:

```
>S
if upsecs%3==0 {
	=>Publish tele/%topic%/SENSOR {"SM":{"Energie Bezug":%2sml[1]%}}
}
```

%2sml[1]%  is the value of sml decoder line 1 with 2 decimal places

add all the other values to the publish cmd above

**Reply by Toto1308** [reply]
Thank you very much for your help.

**Comment by Toto1308** [comment]
![grafik](https://github.com/user-attachments/assets/fbe5053e-a900-44e6-b5f0-c571d6f67fe6)

**Comment by Toto1308** [comment]
My script looks like this now:

`>D
>B
=>teleperiod 30
=>sensor53 r
>M 1
; Device: eBZ DD3 2R06 ODZ1
; protocol is D0 OBIS ASCII
; 9600@7E1 for OP-type devices, 9600@8N1 for SM-type devices
+1,3,o,0,9600,SM,1
; Zählerstand zu +A, tariflos, 
; Zählerstände Auflösung 10 µW*h (6 Vorkomma- und 8 Nachkommastellen)
1,1-0:1.8.0*255(@0.001,Energie Bezug Σ,Wh,1_8_0,18
; Zählerstand zu +A, Tarif 1
1,1-0:1.8.1*255(@0.001,Energie Bezug T1,Wh,1_8_1,18
; Zählerstand zu +A, Tarif 2
1,1-0:1.8.2*255(@0.001,Energie Bezug T2,Wh,1_8_2,18
; Zählerstand zu -A, tariflos
1,1-0:2.8.0*255(@0.001,Energie Export,Wh,2_8_0,18
; Summe der Momentan-Leistungen in allen Phasen, Auflösung 0,01W (5 Vorkomma- und 2 Nachkommastellen)
1,1-0:16.7.0*255(@1,Leistung Σ,W,16_7_0,18
; Momentane Leistung in Phase Lx, Auflösung 0,01W (5 Vorkomma- und 2 Nachkommastellen)
1,1-0:36.7.0*255(@1,Leistung L1,W,36_7_0,18
1,1-0:56.7.0*255(@1,Leistung L2,W,56_7_0,18
1,1-0:76.7.0*255(@1,Leistung L3,W,76_7_0,18
; Spannung in Phase Lx, Auflösung 0,1V (nur über MSB)
1,1-0:32.7.0*255(@1,Spannung L1,V,32_7_0,17
1,1-0:52.7.0*255(@1,Spannung L2,V,52_7_0,17
1,1-0:72.7.0*255(@1,Spannung L3,V,72_7_0,17
; Statuswort, 4 Byte Information über den Betriebszustand, HEX string
; tasmota can decode one string per device only!
;1,1-0:96.5.0*255(@#),Status1,,96_5_0,16
;1,1-0:96.8.0*255(@#),Status2,,96_8_0,16
; Geräte-Identifikation, Nach DIN 43863-5 
1,1-0:96.1.0*255(@#),Identifikation,,96_1_0,16
;1,1-0:0.0.0*255(@#),Identifikation,,0_0_0,16
#`
